apps/schema/src/components/SchemaApp.vue

```diff
@@ -0,0 +1,292 @@
+    };
+  },
+  methods: {
+    handleInput() {
+      this.dirty = true;
+      this.success = null;
+      this.key = Date.now();
+      this.$emit("input", this.schema);
+    },
+    toggleShowDiagram() {
+      this.showDiagram = !this.showDiagram;
+      this.$scrollTo({ el: "#molgenis_diagram_anchor", offset: -50 });
+    },
+    saveSchema() {
+      this.loading = true;
+      this.graphqlError = null;
+      this.warning = "submitting changes";
+      this.success = null;
+      //copy so in case of error user can continue to edit
+      let schema = JSON.parse(JSON.stringify(this.schema));
+      let tables = schema.tables ? schema.tables : [];
+
+      //transform subclasses back into their original tables.
+      //create a map of tables
+      const tableMap = tables.reduce((map, table) => {
+        map[table.name] = table;
+        if (table.subclasses) {
+          table.subclasses.forEach(
+            (subclass) => (map[subclass.name] = subclass)
+          );
+        }
+        delete table.subclasses;
+        return map;
+      }, {});
+      //redistribute the columns to subclasses
+      tables.forEach((table) => {
+        if (table.columns !== undefined) {
+          table.columns.forEach((column) => {
+            if (column.table !== table.name) {
+              tableMap[column.table].columns.push(column);
+            }
+          });
+        }
+      });
+      tables.forEach((table) => {
+        table.columns = table.columns
+          ? table.columns.filter((column) => column.table === table.name)
+          : [];
+      });
+      tables = Object.values(tableMap);
+      //add ontologies
+      tables.push(...schema.ontologies);
+      request(
+        "graphql",
+        `mutation change($tables:[MolgenisTableInput]){change(tables:$tables){message}}`,
+        {
+          tables: tables,
+        }
+      )
+        .then(() => {
+          this.warning = "submission complete, reloading schema";
+          this.loadSchema();
+          this.warning = null;
+          this.success = `Schema saved`;
+        })
+        .catch((error) => {
+          if (error.response.status === "403") {
+            this.graphqlError = "Forbidden. Do you need to login?";
+            this.showLogin = true;
+          } else {
+            this.graphqlError = error.response.errors[0].message;
+          }
+          this.warning = null;
+          this.loading = false;
+        });
+      this.loading = false;
+    },
+    loadSchema() {
+      console.log("load schema");
+      this.graphqlError = null;
+      this.loading = true;
+      request(
+        "graphql",
+        "{_session{schemas,roles}_schema{name,tables{name,tableType,inherit,externalSchema,description,semantics,columns{name,table,position,columnType,inherited,key,refSchema,refTable,refLink,refBack,required,description,semantics,validation,visible}}}}"
+      )
+        .then((data) => {
+          let _schema = this.addOldNamesAndRemoveMeta(data._schema);
```

Comment:
const , avoid passing a object to a function that changes the object

Reply:
agreed